vdaf/prio3/arith/arith.go

```diff
@@ -0,0 +1,133 @@
+//go:generate go run gen.go
+
+// Package arith provides arithmetic operations over prime fields, vectors,
+// and polynomials.
+package arith
+
+import (
+	"encoding"
+	"io"
+
+	"github.com/cloudflare/circl/internal/conv"
+	"github.com/cloudflare/circl/internal/sha3"
+	"golang.org/x/crypto/cryptobyte"
+)
+
+// Elt is any type that stores a prime field element.
+type Elt any
+
+// Fp lists the functionality that a prime field element must have denoting
+// methods with a pointer receiver.
+type Fp[E Elt] interface {
+	*E
+	// Size returns the number of bytes to encode a field element.
+	Size() uint
+	// Returns true if the element is the neutral additive element.
+	IsZero() bool
+	// Returns true if the element is the neutral multiplicative element.
+	IsOne() bool
+	// Returns true if the element is equivalent to x.
+	IsEqual(x *E) bool
+	// Set the element to the neutral additive element.
+	SetOne()
+	// Set the element to x if x < Order().
+	SetUint64(uint64) error
+	// Returns the integer representative of the element if x < 2^64.
+	GetUint64() (x uint64, err error)
+	// Set the element to the principal root of unity of order 2^n.
+	SetRootOfUnityTwoN(n uint)
+	// AddAssing calculates z = z + x.
+	AddAssign(x *E)
+	// SubAssign calculates z = z - x.
+	SubAssign(x *E)
+	// MulAssign calculates z = z * x.
+	MulAssign(x *E)
+	// Add calculates z = x + y.
+	Add(x, y *E)
+	// Sub calculates z = x - y.
+	Sub(x, y *E)
+	// Mul calculates z = x * y.
+	Mul(x, y *E)
+	// Sqr calculates z = x * x.
+	Sqr(x *E)
+	// Inv calculates z = 1 / x.
+	Inv(x *E)
+	// InvUint64 calculates z = 1 / x.
+	InvUint64(x uint64)
+	// InvTwoN calculates z = 1 / 2^n.
+	InvTwoN(x uint)
+	// Random samples an element from an io.Reader.
+	Random(io.Reader) error
+	// RandomSHA3 samples an element from a SHA3 state.
+	RandomSHA3(*sha3.State) error
+	// Encodes an element to bytes.
+	encoding.BinaryMarshaler
+	// Decodes an element from bytes.
+	encoding.BinaryUnmarshaler
+	// Encodes an element using a cryptobyte.Builder.
+	cryptobyte.MarshalingValue
+	// Decodes an element from a cryptobyte.String.
+	conv.UnmarshalingValue
```

Comment:
one uses the other, `cryptobyte` methods are used since they implement the TLS syntax, while `encoding` methods are (most of the times) just calls to the previous ones.